### PR TITLE
Average precision score

### PR DESCRIPTION
**Pull request by btrotta**
Implement average precision score for binary classification. This is very similar to the precision-recall AUC requested in #3026, but it does not interpolate the points of the curve, since this is not recommended for PR curves (see https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html).

### CODE REVIEW

File: docs/Parameters.rst
Comment (jameslamb) [inline docs/Parameters.rst:1013]: could this be called `average_precision` instead of `ap`? I think the very short names can be confusing unless they're very common acronyms like AUC.

Comment (btrotta) [reply docs/Parameters.rst]: Done :)

Comment (jameslamb) [reply docs/Parameters.rst]: awesome, thanks! I think it will help

**Review by jameslamb — APPROVED** [review]
looks good to me from the R side! I think another reviewer should approve before this is merged though.

**Comment by jameslamb** [comment]
@btrotta in case you aren't aware of the practice...I just added the `feature` label to this so it will be categorized correctly in the next release: https://github.com/microsoft/LightGBM/releases/tag/untagged-bc49d5d1a6b7637e4d22

this is done automatically based on tags by https://github.com/microsoft/LightGBM/blob/master/.github/release-drafter.yml

**Comment by btrotta** [comment]
@jameslamb  Thanks! I didn't realize the tags were used for the release notes.

**Comment by jameslamb** [comment]
no problem!

**Review by guolinke — APPROVED** [review]

File: src/metric/binary_metric.hpp
Comment (StrikerRUS) [inline src/metric/binary_metric.hpp:287]: @btrotta 
Was name left `ap` (instead of `average_precision`, https://github.com/microsoft/LightGBM/pull/3347#discussion_r482112568) intentionally?

Comment (btrotta) [reply src/metric/binary_metric.hpp]: Oops, this was a mistake, fixed now.

**Pull request merged**
